File: basic_dsa_one_hundred_eighty_two.java
/**
 * Basic DSA Problems - Koko Eating Bananas
 * Min eating speed so all piles are eaten in h hours (each hour one pile, k bananas per hour from that pile)
 */
public class basic_dsa_one_hundred_eighty_two {

    // Problem: piles[i] = bananas in pile i. In h hours, finish all. k = bananas per hour per pile. Find min k.

    /**
     * Binary search on k. For each k, hours = sum(ceil(pile/k)). If hours <= h, try smaller k.
     * Time O(n log max), Space O(1)
     */
    public static int minEatingSpeed(int[] piles, int h) {
        int lo = 1, hi = 0;
        for (int p : piles) if (p > hi) hi = p;
        while (lo < hi) {
            int k = lo + (hi - lo) / 2;
            long hours = 0;
            for (int p : piles) hours += (p + k - 1) / k;
            if (hours <= h) hi = k;
            else lo = k + 1;
        }
        return lo;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Koko Eating Bananas");
        System.out.println("=".repeat(60));

        System.out.println("[3,6,7,11], h=8 -> " + minEatingSpeed(new int[]{3, 6, 7, 11}, 8));
        System.out.println("[30,11,23,4,20], h=5 -> " + minEatingSpeed(new int[]{30, 11, 23, 4, 20}, 5));
    }
}

File: basic_dsa_one_hundred_eighty_two.py
"""
Basic DSA - Find Smallest Letter Greater Than Target
Given a sorted list of letters and a target letter, return the smallest letter in
the list that is strictly greater than target. Letters wrap around.
"""


def next_greatest_letter(letters, target):
    """
    Binary search for first letter > target. Wrap around using modulo.
    Time Complexity: O(log n)
    Space Complexity: O(1)
    """
    lo, hi = 0, len(letters)
    while lo < hi:
        mid = (lo + hi) // 2
        if letters[mid] <= target:
            lo = mid + 1
        else:
            hi = mid
    return letters[lo % len(letters)]


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Find Smallest Letter Greater Than Target")
    print("=" * 60)

    print(next_greatest_letter(["c", "f", "j"], "a"))  # c
    print(next_greatest_letter(["c", "f", "j"], "c"))  # f
    print(next_greatest_letter(["c", "f", "j"], "d"))  # f
    print(next_greatest_letter(["c", "f", "j"], "j"))  # c
